JS/Lesson20/emrah/exercises.js

```diff
@@ -1,70 +1,132 @@
+const triangle = new Shape("triangle", "Blue");
+const square = new Square("Blue", 5);
 
 console.log(triangle.describe());
 console.log(square.describe());
 console.log(square.area());
 
 console.log(square);
 
+// 3. Create a Vehicle class with make, model, and year properties, and a start() method that logs a message to the console indicating that the vehicle has started. Then create a Car class that extends the Vehicle class and adds a numDoors property. Override the start() method in the Car class to log a different message to the console indicating that the car has started.
 
+class Vehicle {
+  constructor(make, model, year) {
+    this._make = make;
+    this._model = model;
+    this._year = year;
+  }
+  start() {
+    console.log("The vehicle has started");
+  }
+}
+class Car extends Vehicle {
+  constructor(make, model, year, numDoors) {
+    super(make, model, year);
+    this._numDoors = numDoors;
+  }
+  start() {
+    console.log("The car has started");
+  }
+}
 
-
-// 3. Create a Vehicle class with make, model, and year properties, and a start() method that logs a message to the console indicating that the vehicle has started. Then create a Car class that extends the Vehicle class and adds a numDoors property. Override the start() method in the Car class to log a different message to the console indicating that the car has started.
+const myCar = new Car("Nissan", "Qashqai", 2022, 4);
+myCar.start();
 
 // 4. Create a BankAccount class with balance and interestRate properties, and deposit() and withdraw() methods. Add a transactionHistory property that is an array of objects representing each transaction made on the account, with properties for type (either "deposit" or "withdrawal"), amount, and date ({type: "withdrawal", amount: amount, date: new Date().toLocaleDateString()}). Implement deposit() and withdraw() methods that add new transactions to the transactionHistory array. Implement a get transactionHistory() getter method that returns a copy of the transaction history array, and a get currentBalance() getter method that calculates and returns the current balance of the account based on the transaction history.
+
+class BankAccount {
+  constructor(balance, interestRate) {
+    this._balance = balance;
+    this._interestRate = interestRate;
+    this.transactionHistory = [];
+  }
+  deposit(amount) {
+    this._amount += amount;
+    this.transactionHistory.push({
+      type: "deposit",
+      amount: amount,
+      date: new Date().toLocaleDateString(),
+    });
+  }
+  withdraw(amount) {
+    if (amount <= this.balance) {
+      this._balance -= amount;
+      this.transactionHistory.push({
+        type: "withdrawal",
+        amount: amount,
+        date: new Date().toLocaleDateString(),
+      });
+    } else {
+      console.log("Insufficient funds");
+    }
+  }
+  get transactionHistory() {
+    return this.transactionHistory.slice();
+  }
+  get currentBalance() {
+    let balance = this._balance;
```

Comment:
Since you update balance in both deposit and withdraw operations, you can just return it (`return this._balance;`). No need to do anything more.